### PR TITLE
confignetwork configure bridge using nmcli 

### PR DESCRIPTION
**Pull request by bybai**
### The PR is for task 
xcat2/xcat2-task-management#611

### The content of the PR:
configure bridge scripts using nmcli

### The UT result
1. create bridge using nmcli, when there is no old connection existed
```
]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: br51 ens5
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : br51 ens5
byrh09: [I]: create_bridge_interface_nmcli ifname=br51 _brtype=bridge _port=ens5 _pretype=ethernet _ipaddr=50.4.41.141
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "br51".
byrh09: [I]: create bridge connection xcatbr51
byrh09: Connection 'xcatbr51' (6041e379-6a6a-44e0-984f-3524f150270f) successfully added.
byrh09: [I]: create ethernet slaves connetcion xcat_br_ens5 for bridge
byrh09: Connection 'xcat_br_ens5' (39bc4779-2b88-470a-9553-3d08178576a4) successfully added.
byrh09: [I]: add ip 50.4.41.141/8 to bridge
byrh09: [I]: nmcli con up xcat_br_ens5; nmcli con up xcatbr51
byrh09: [I]: State of "br51" was "DOWN" instead of expected "UP". Wait 0 of 20 with interval 40.
byrh09: [I]: [bridge] >> 48: br51: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
byrh09: [I]: [bridge] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: [I]: [bridge] >>     inet 50.4.41.141/8 brd 50.255.255.255 scope global noprefixroute br51
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: [I]: [bridge] >>     inet6 fe80::1fa8:e5d5:94bd:c932/64 scope link noprefixroute
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: postscript end....: confignetwork exited with code 0
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```
On compute node
```
]# ip address show dev br51
48: br51: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
    inet 50.4.41.141/8 brd 50.255.255.255 scope global noprefixroute br51
       valid_lft forever preferred_lft forever
    inet6 fe80::1fa8:e5d5:94bd:c932/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
```

2. create bridge using nmcli, when there is old connection existed
```
]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: br51 ens5
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : br51 ens5
byrh09: [I]: create_bridge_interface_nmcli ifname=br51 _brtype=bridge _port=ens5 _pretype=ethernet _ipaddr=50.4.41.141
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "br51".
byrh09: [I]: xcatbr51 exists, rename old xcatbr51 to xcatbr51-tmp
byrh09: [I]: create bridge connection xcatbr51
byrh09: Connection 'xcatbr51' (0a652fb5-03a3-4c07-af48-dbf479c79599) successfully added.
byrh09: [I]: xcat_br_ens5 exists, rename old connetion xcat_br_ens5 to xcat_br_ens5-tmp
byrh09: [I]: create ethernet slaves connetcion xcat_br_ens5 for bridge
byrh09: Connection 'xcat_br_ens5' (a2c03268-ef47-4965-a04e-24f882a84651) successfully added.
byrh09: [I]: add ip 50.4.41.141/8 to bridge
byrh09: [I]: nmcli con up xcat_br_ens5; nmcli con up xcatbr51
byrh09: Connection 'xcatbr51-tmp' (6041e379-6a6a-44e0-984f-3524f150270f) successfully deleted.
byrh09: Connection 'xcat_br_ens5-tmp' (39bc4779-2b88-470a-9553-3d08178576a4) successfully deleted.
byrh09: [I]: State of "br51" was "DOWN" instead of expected "UP". Wait 0 of 20 with interval 40.
byrh09: [I]: [bridge] >> 48: br51: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
byrh09: [I]: [bridge] >>     link/ether 42:9b:0a:04:29:09 brd ff:ff:ff:ff:ff:ff
byrh09: [I]: [bridge] >>     inet 50.4.41.141/8 brd 50.255.255.255 scope global noprefixroute br51
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: [I]: [bridge] >>     inet6 fe80::300e:ca91:d38e:807/64 scope link noprefixroute
byrh09: [I]: [bridge] >>        valid_lft forever preferred_lft forever
byrh09: postscript end....: confignetwork exited with code 0
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```

3. network interface does not exists
```
]# updatenode byrh09 confignetwork
byrh09: =============updatenode starting====================
byrh09: trying to download postscripts...
byrh09: postscripts downloaded successfully
byrh09: trying to get mypostscript from 10.4.41.1...
byrh09: postscript start..: confignetwork
byrh09: [I]: NetworkManager is active
byrh09: [I]: All valid nics and device list:
byrh09: [I]: br51 ens6
byrh09: ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
byrh09: configure nic and its device : br51 ens6
byrh09: [I]: create_bridge_interface_nmcli ifname=br51 _brtype=bridge _port=ens6 _pretype=ethernet _ipaddr=50.4.41.141
byrh09: [I]: Pickup xcatnet, "50_0_0_0-255_0_0_0", from NICNETWORKS for interface "br51".
byrh09: [E]:Error: Device ens6 not found
byrh09: postscript end....: confignetwork exited with code 1
byrh09: Running of postscripts has completed.
byrh09: =============updatenode ending====================
```

### CODE REVIEW

**Comment by bybai** [comment]
I am OK to merge it. Thanks @zet809

File: xCAT/postscripts/nicutils.sh [hunk 1]
```diff
@@ -1618,12 +1618,165 @@ function create_vlan_interface_nmcli {
...
+        log_info "create bridge connection $xcat_con_name"
+        $nmcli con add type bridge con-name $xcat_con_name ifname $ifname
+        if [ $? -ne 0 ]; then
+            log_error "nmcli failed to add bridge $ifname"
+            return 1
+        fi
+    elif [ x"$_brtype" == "xbridge_ovs" ]; then
+        log_error "OVSBridge is not supported."
+        return 1
+    fi
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: What happened if brtype other than "brideg" and "bridge_ovs"?

File: xCAT/postscripts/nicutils.sh [hunk 2]
```diff
@@ -1618,12 +1618,165 @@ function create_vlan_interface_nmcli {
...
+            log_info "$xcat_con_name exists, rename old $xcat_con_name to $tmp_con_name"
+            $nmcli con modify $xcat_con_name connection.id $tmp_con_name
+            if [ $? -ne 0 ] ; then
+                log_error "$nmcli rename $xcat_con_name failed"
+                return 1
+            fi
+        fi
+        log_info "create bridge connection $xcat_con_name"
+        $nmcli con add type bridge con-name $xcat_con_name ifname $ifname
+        if [ $? -ne 0 ]; then
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: Need to modify the con_name back if failed.

File: xCAT/postscripts/nicutils.sh [hunk 3]
```diff
@@ -1618,12 +1618,165 @@ function create_vlan_interface_nmcli {
...
+            $nmcli con modify $xcat_slave_con connection.id $tmp_slave_con_name
+            if [ $? -ne 0 ] ; then
+                log_error "$nmcli rename $xcat_slave_con failed"
+                return 1
+            fi
+        fi
+        log_info "create $_pretype slaves connetcion $xcat_slave_con for bridge"
+        $nmcli con add type $_pretype con-name $xcat_slave_con ifname $_port master $ifname;
+        if [ $? -ne 0 ]; then
+            log_error "nmcli failed to add bridge slave $_port"
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: Shall rename the tmp_slav_con_name back if failed.

File: xCAT/postscripts/nicutils.sh [hunk 4]
```diff
@@ -1618,12 +1618,165 @@ function create_vlan_interface_nmcli {
...
+    log_info "$nmcli con up $xcat_slave_con; $nmcli con up $xcat_con_name"
+    lines=`$nmcli con up $xcat_slave_con; $nmcli con up $xcat_con_name`
+    rc=$?
+
+    # If bridge interface is active, delete tmp old connection
+    # If bridge interface is not active, delete new bridge and slave connection, and restore old connection
+    is_connection_activate_intime $xcat_con_name
+    is_active=$?
+    if [ "$is_active" -eq 0 ]; then
+        log_error "$nmcli con up $xcat_con_name failed with return code equals to $rc"
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: rc? shall be is_active

File: xCAT/postscripts/nicutils.sh [hunk 5]
```diff
@@ -1618,12 +1618,165 @@ function create_vlan_interface_nmcli {
...
+        $nmcli con delete $xcat_con_name
+        is_nmcli_connection_exist $tmp_con_name
+        if [ $? -eq 0 ]; then
+            nmcli con modify $tmp_con_name connection.id $xcat_con_name
+        fi
+        $nmcli con delete $xcat_slave_con
+        is_nmcli_connection_exist $tmp_slave_con_name
+        if [ $? -eq 0 ]; then
+            nmcli con modify $xcat_slave_con connection.id $tmp_slave_con_name
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: shall be: `nmcli con modify $tmp_slave_con_name connection.id $xcat_slave_con`

File: xCAT/postscripts/confignetwork [hunk 6]
```diff
@@ -562,7 +561,11 @@ function configure_nicdevice {
                if [ "$networkmanager_active" = "0" ]; then
                    create_bridge_interface ifname=$nic_dev _brtype=$nic_dev_type _port=$base_nic_dev _pretype=$base_nic_type
                elif [ "$networkmanager_active" = "1" ]; then
-                   create_bridge_interface_nmcli ifname=$nic_dev _brtype=$nic_dev_type _port=$base_nic_dev _pretype=$base_nic_type
+                   create_bridge_interface_nmcli ifname=$nic_dev _brtype=$nic_dev_type _port=$base_nic_dev _pretype=$base_nic_type _ipaddr=$ipaddrs
+                   if [ $? -ne 0 ]; then
+                       log_error "create bridge interface $ifname failed"
```
Comment (zet809) [inline xCAT/postscripts/confignetwork]: Shall be nic_dev right? There is no $ifname

File: xCAT/postscripts/nicutils.sh [hunk 7]
```diff
@@ -1618,12 +1618,175 @@ function create_vlan_interface_nmcli {
...
+           [ "$key" = "_brtype" ] || \
+           [ "$key" = "_pretype" ] || \
+           [ "$key" = "_port" ] || \
+           [ "$key" = "_ipaddr" ]; then
+            eval "$1"
+        fi
+        shift
+    done
+    # query "nicnetworks" table about its target "xcatnet"
+    if [ -n "$ifname" ]; then
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: No need to check ifname, right?

File: xCAT/postscripts/nicutils.sh [hunk 8]
```diff
@@ -1618,12 +1618,175 @@ function create_vlan_interface_nmcli {
...
+        fi
+        log_info "create bridge connection $xcat_con_name"
+        $nmcli con add type bridge con-name $xcat_con_name ifname $ifname
+        if [ $? -ne 0 ]; then
+            log_error "nmcli failed to add bridge $ifname"
+            is_nmcli_connection_exist $tmp_con_name
+            if [ $? -eq 0 ] ; then
+                $nmcli con modify $tmp_con_name connection.id $xcat_con_name
+            fi
+            $nmcli con delete $xcat_con_name
```
Comment (zet809) [inline xCAT/postscripts/nicutils.sh]: Shall not delete the connection here, since it is the worked connection.

**Comment by zet809** [comment]
Ok to me, thx @bybai

**Pull request merged**